docs/interfaces.md

```diff
@@ -0,0 +1,431 @@
+
+```go
+type EnclaveChaincode struct {
+        erccStub ercc.EnclaveRegistryStub
+        tlccStub tlcc.TLCCStub
+        enclave  enclave.Stub
+        verifier crypto.Verifier
+        ev       attestation.EvidenceService // abstracts evidence creation for EPID/DCAP
+}
+
+type StubImpl struct {
+	eid C.enclave_id_t
+}
+```
+
+
+# ECC_Enclave
+
+## Interface:
+The ECC_Enclave interface specifies interface of an FPC chaincode enclave. 
+
+Note: `sealed credentials` is renamed to `sealed state`
+
+```c++
+// creates an new enclave
+public int ecall_create_enclave(
+        [in, size=cc_params_len] uint8_t *cc_params, uint32_t cc_params_len,
+        [out] uint8_t *credentials,     // TBD is this a fixed length type?
+        [out, size=max_sealed_state_len] uint8_t *sealed_state, uint32_t max_sealed_state_len,
+        [out] uint32_t *sealed_data_len);
+
+// creates ("restores") an enclave with sealed state
+public int ecall_create_enclave_with_sealed_state(
+        [in, size=sealed_state_len] uint8_t *sealed_state, uint32_t sealed_state_len);
+
+// TBD what is the input/output here?
+public int ecall_get_CSR(
+        [in]    
+        [out] csr);
+
+// invoke a FPC chaincode
+public int ecall_chaincode_invoke(
+        [in, size=proposal_len]  uint8_t *proposal, uint32_t proposal_len,
+        [out, size=response_len_in] uint8_t *response, uint32_t response_len_in,
+        [out] uint32_t *response_len_out,
+        [out] sgx_ec256_signature_t *signature,
+        [user_check] void *u_shim_ctx);
+
+// key generation and distribution
+// TODO interface below needs more brain cycles
+public int ecall_generate_cc_keys(
+        // TBD: Why do we need cc_params?
+        [in, size=cc_params_len] uint8_t *cc_params, uint32_t cc_params_len, 
+        [out] uint8_t *cckey_registration_msg,
+        [out] uint8_t *sealed_cckeys);
+
+public int ecall_export_cc_keys(
+       [in] uint8_t *target_enclave_vk,
+       [out] uint8_t *export_msg;
+
+public int ecall_import_cc_keys(
+        // TBD: Why do we need cc_params?
+        [in, size=cc_params_len] uint8_t *cc_params, uint32_t cc_params_len, 
+        [in] uint8_t *export_msg,
+        [out] uint8_t *cckey_registration_msg,
+        [out] uint8_t *sealed_cckeys);
+```
+
+## State:
+We also use the state, in particular, `internal_ecc_state` to control that `ecall_create_enclave` or `ecall_create_enclave_with_sealed_credentials` can be invoked once.
+
+TODO: change key types to whatever pdo crypto gives us.
+```c++
+
+// enclave input on create; this is kept inside the enclave
+typedef struct cc_params_t {
+        cc_id,          // chaincode id
+        version,        // chaincode version, this is the expected mrenclave
+        msp_id,         // msp_id that does what?
+        channel_id      // channel id
+        tlcc_mrenclave, // expected tlcc mrenclave
+};
+
+// this is create on enclave creation and kept inside the enclave
+typedef struct internal_ecc_state_t {
```

Comment:
Maybe worth from to separate the sealed vs ephemeral state?  E.g., session key definitely would not be sealed.  CC_param certainly has to be bound to sealed state but it could actually be beneficial to bind it via authenticated data as the untrusted state anyway needs mapping of sealed blob to a subset of CC_Param?)   That said, might be driven also by the API of the seal/unseal library (if we go a modular route)?

Reply:
also, might be worth adding explicit an `initialization_state` enum or alike which with potential states `uninitialized`, 'enclave_keys_exists`, `cc_key_exists` (after which probably invoke should work? Or would we also need a `committed` state or alike?)